src/AutorestSwift/Models/Complex/ParameterType.swift

```diff
@@ -0,0 +1,115 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import Foundation
+
+enum ParameterType: Codable {
+    case regular(Parameter)
+    case virtual(VirtualParameter)
+
+    // MARK: Codable
+
+    init(from decoder: Decoder) throws {
+        if let decoded = try? VirtualParameter(from: decoder) {
+            self = .virtual(decoded)
+        } else if let decoded = try? Parameter(from: decoder) {
+            self = .regular(decoded)
+        } else {
+            fatalError("Unable to decode ParameterType")
+        }
+    }
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.unkeyedContainer()
+        switch self {
+        case let .regular(param):
+            try container.encode(param)
```

Comment:
since both cases are doing the same thing, do we need a switch statement here?

Reply:
It seems so. While it does look like they are doing the same thing, they actually aren't.

```Swift
        switch self {
        case let .regular(param):
            // encodes a Parameter
            try container.encode(param)
        case let .virtual(param):
            // encodes a VirtualParamter
            try container.encode(param)
        }
```

I tried to just `container.encode(self)` and it caused and infinite loop and stack overflow. This is basically analogous to the logic we used elsewhere of trying to decode as one type then the next, etc.